pkg/iter/iterator.go

```diff
@@ -430,3 +431,98 @@ func (i *entryIteratorBackward) Error() error { return nil }
 func (i *entryIteratorBackward) Labels() string {
 	return ""
 }
+
+type logentry struct {
+	entry  logproto.Entry
+	labels string
+}
+
+type peekingEntryIterator struct {
+	iter EntryIterator
+
+	cache *logentry
+	next  *logentry
+}
+
+// PeekingEntryIterator is an entry iterator that can look ahead an entry
+// using `Peek` without advancing its cursor.
+type PeekingEntryIterator interface {
```

Comment:
I need to change this with @sandlis branch, he has a better approach with the heap version.

Reply:
see https://github.com/grafana/loki/pull/541/files#diff-5cf7837c33ac75c4bf2be956919db224R103

Reply:
Actually @sandlis I'm not sure the `HeapIterator.Peek` gives you the next entry, it doesn't seems to account yet for deduplication.

Reply:
@sandlis, @gouthamve  I'm curious to know what you think about this. I feel like I need to keep this one.

Reply:
HeapIterator.Peek does not return whole entry, it just returns next max/min timestamp. I guess you are looking for 2 prefetched entries here while existing implementation only resolves and keeps single entry. If we want to keep this I would suggest peekingEntryIterator.Next() should return true until both the entries are consumed. Right now it returns false when cacheNext() returns false. With this change, Peek() could return nil values when the cache is nil. This is to make sure it behaves same as other iterators and the last entry not getting missed because someone missed getting the entries after the peekingEntryIterator.Next() loop. Just my thoughts about it.

Reply:
make sense will adjust

Reply:
done